File: packages/spark-sdk/ios/SparkTokenPrimitivesModule.swift
import Foundation
import React

@objc(SparkTokenPrimitivesModule)
class SparkTokenPrimitivesModule: NSObject, RCTBridgeModule {

    @objc
    static func moduleName() -> String! {
        return "SparkTokenPrimitivesModule"
    }

    @objc
    static func requiresMainQueueSetup() -> Bool {
        return false
    }

    private func arrayToData(_ array: [Any]) -> Data? {
        return (array as? [Int])?.map { UInt8($0) }.data
    }

    private func optionalArrayToData(_ value: Any?) -> Data? {
        guard let array = value as? [Any] else { return nil }
        return arrayToData(array)
    }

    private func dataToArray(_ data: Data) -> [Int] {
        return Array(data).map { Int($0) }
    }

    private func parseSelectedOutput(_ dict: [String: Any]) throws -> SelectedTokenOutput {
        guard let prevHashArray = dict["previousTransactionHash"] as? [Any],
              let previousTransactionHash = arrayToData(prevHashArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid previousTransactionHash format"])
        }
        guard let voutInt = dict["previousTransactionVout"] as? Int else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid previousTransactionVout format"])
        }
        guard let ownerArray = dict["ownerPublicKey"] as? [Any],
              let ownerPublicKey = arrayToData(ownerArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid ownerPublicKey format"])
        }
        guard let tokenIdArray = dict["tokenIdentifier"] as? [Any],
              let tokenIdentifier = arrayToData(tokenIdArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid tokenIdentifier format"])
        }
        guard let tokenAmountArray = dict["tokenAmount"] as? [Any],
              let tokenAmount = arrayToData(tokenAmountArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid tokenAmount format"])
        }
        return SelectedTokenOutput(
            previousTransactionHash: previousTransactionHash,
            previousTransactionVout: UInt32(voutInt),
            ownerPublicKey: ownerPublicKey,
            tokenIdentifier: tokenIdentifier,
            tokenAmount: tokenAmount
        )
    }

    private func parseReceiverOutput(_ dict: [String: Any]) throws -> ReceiverTokenOutput {
        guard let receiverSparkAddress = dict["receiverSparkAddress"] as? String else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid receiverSparkAddress format"])
        }
        let tokenIdentifier = optionalArrayToData(dict["tokenIdentifier"])
        let tokenAmount = optionalArrayToData(dict["tokenAmount"])
        return ReceiverTokenOutput(
            receiverSparkAddress: receiverSparkAddress,
            tokenIdentifier: tokenIdentifier,
            tokenAmount: tokenAmount
        )
    }

    private func parseSignatureWithIndex(_ dict: [String: Any]) throws -> SignatureWithIndexInput {
        guard let inputIndex = dict["inputIndex"] as? Int else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid inputIndex format"])
        }
        guard let publicKeyArray = dict["publicKey"] as? [Any],
              let publicKey = arrayToData(publicKeyArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid publicKey format"])
        }
        guard let signatureArray = dict["signature"] as? [Any],
              let signature = arrayToData(signatureArray) else {
            throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid signature format"])
        }
        return SignatureWithIndexInput(
            inputIndex: UInt32(inputIndex),
            publicKey: publicKey,
            signature: signature
        )
    }

    @objc(constructPartialTransferTransaction:resolve:reject:)
    func rn_constructPartialTransferTransaction(_ params: [String: Any],
                                                resolve: @escaping RCTPromiseResolveBlock,
                                                reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let identityPublicKeyArray = params["identityPublicKey"] as? [Any],
                  let identityPublicKey = arrayToData(identityPublicKeyArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid identityPublicKey format"])
            }

            guard let selectedOutputsArray = params["selectedOutputs"] as? [[String: Any]] else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid selectedOutputs format"])
            }
            let selectedOutputs = try selectedOutputsArray.map { try parseSelectedOutput($0) }

            guard let receiverOutputsArray = params["receiverOutputs"] as? [[String: Any]] else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid receiverOutputs format"])
            }
            let receiverOutputs = try receiverOutputsArray.map { try parseReceiverOutput($0) }

            guard let operatorKeysArray = params["operatorIdentityPublicKeys"] as? [[Any]] else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid operatorIdentityPublicKeys format"])
            }
            let operatorIdentityPublicKeys: [Data] = try operatorKeysArray.enumerated().map { (i, arr) in
                guard let data = arrayToData(arr) else {
                    throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid operator key at index \(i)"])
                }
                return data
            }

            guard let network = params["network"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid network format"])
            }
            guard let validityDurationSeconds = params["validityDurationSeconds"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid validityDurationSeconds format"])
            }
            guard let clientCreatedTimestampUnixMicros = params["clientCreatedTimestampUnixMicros"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid clientCreatedTimestampUnixMicros format"])
            }
            guard let withdrawBondSats = params["withdrawBondSats"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid withdrawBondSats format"])
            }
            guard let withdrawRelativeBlockLocktime = params["withdrawRelativeBlockLocktime"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid withdrawRelativeBlockLocktime format"])
            }

            let executeBeforeUnixMicros: Int64?
            if let raw = params["executeBeforeUnixMicros"] as? Int {
                executeBeforeUnixMicros = Int64(raw)
            } else {
                executeBeforeUnixMicros = nil
            }

            let request = TransferBuildRequest(
                identityPublicKey: identityPublicKey,
                selectedOutputs: selectedOutputs,
                receiverOutputs: receiverOutputs,
                operatorIdentityPublicKeys: operatorIdentityPublicKeys,
                network: UInt32(network),
                validityDurationSeconds: UInt64(validityDurationSeconds),
                clientCreatedTimestampUnixMicros: Int64(clientCreatedTimestampUnixMicros),
                withdrawBondSats: UInt64(withdrawBondSats),
                withdrawRelativeBlockLocktime: UInt64(withdrawRelativeBlockLocktime),
                executeBeforeUnixMicros: executeBeforeUnixMicros
            )

            let result = try constructPartialTransferTransaction(request: request)

            let resultDict: [String: Any] = [
                "partialTokenTransactionBytes": dataToArray(result.partialTokenTransactionBytes),
                "partialTokenTransactionHash": dataToArray(result.partialTokenTransactionHash)
            ]
            resolve(resultDict)
        } catch {
            reject("ERROR_CONSTRUCT_PARTIAL_TRANSFER_TRANSACTION", error.localizedDescription, error)
        }
    }

    @objc(hashPartialTokenTransaction:resolve:reject:)
    func rn_hashPartialTokenTransaction(_ params: [String: Any],
                                        resolve: @escaping RCTPromiseResolveBlock,
                                        reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let bytesArray = params["partialTokenTransactionBytes"] as? [Any],
                  let bytes = arrayToData(bytesArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid partialTokenTransactionBytes format"])
            }
            let result = try hashPartialTokenTransaction(partialTokenTransactionBytes: bytes)
            resolve(dataToArray(result))
        } catch {
            reject("ERROR_HASH_PARTIAL_TOKEN_TRANSACTION", error.localizedDescription, error)
        }
    }

    @objc(buildBroadcastTransactionRequest:resolve:reject:)
    func rn_buildBroadcastTransactionRequest(_ params: [String: Any],
                                             resolve: @escaping RCTPromiseResolveBlock,
                                             reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let identityPublicKeyArray = params["identityPublicKey"] as? [Any],
                  let identityPublicKey = arrayToData(identityPublicKeyArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid identityPublicKey format"])
            }
            guard let bytesArray = params["partialTokenTransactionBytes"] as? [Any],
                  let partialTokenTransactionBytes = arrayToData(bytesArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid partialTokenTransactionBytes format"])
            }
            guard let signaturesArray = params["ownerSignatures"] as? [[String: Any]] else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid ownerSignatures format"])
            }
            let ownerSignatures = try signaturesArray.map { try parseSignatureWithIndex($0) }

            let request = BroadcastBuildRequest(
                identityPublicKey: identityPublicKey,
                partialTokenTransactionBytes: partialTokenTransactionBytes,
                ownerSignatures: ownerSignatures
            )

            let result = try buildBroadcastTransactionRequest(request: request)
            resolve(dataToArray(result))
        } catch {
            reject("ERROR_BUILD_BROADCAST_TRANSACTION_REQUEST", error.localizedDescription, error)
        }
    }

    @objc(prepareTokenInvoice:resolve:reject:)
    func rn_prepareTokenInvoice(_ params: [String: Any],
                                resolve: @escaping RCTPromiseResolveBlock,
                                reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let receiverIdentityPublicKeyArray = params["receiverIdentityPublicKey"] as? [Any],
                  let receiverIdentityPublicKey = arrayToData(receiverIdentityPublicKeyArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid receiverIdentityPublicKey format"])
            }
            guard let network = params["network"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid network format"])
            }
            let tokenIdentifier = optionalArrayToData(params["tokenIdentifier"])
            let tokenAmount = optionalArrayToData(params["tokenAmount"])
            let memo = params["memo"] as? String
            let senderSparkAddress = params["senderSparkAddress"] as? String
            let expiryTimeUnixMillis: UInt64?
            if let raw = params["expiryTimeUnixMillis"] as? Int {
                expiryTimeUnixMillis = UInt64(raw)
            } else {
                expiryTimeUnixMillis = nil
            }
            let invoiceId = optionalArrayToData(params["invoiceId"])

            let request = PrepareTokenInvoiceRequest(
                receiverIdentityPublicKey: receiverIdentityPublicKey,
                network: UInt32(network),
                tokenIdentifier: tokenIdentifier,
                tokenAmount: tokenAmount,
                memo: memo,
                senderSparkAddress: senderSparkAddress,
                expiryTimeUnixMillis: expiryTimeUnixMillis,
                invoiceId: invoiceId
            )

            let result = try prepareTokenInvoice(request: request)

            let resultDict: [String: Any] = [
                "sparkInvoiceFieldsBytes": dataToArray(result.sparkInvoiceFieldsBytes),
                "sparkInvoiceHash": dataToArray(result.sparkInvoiceHash),
                "unsignedSparkAddress": result.unsignedSparkAddress
            ]
            resolve(resultDict)
        } catch {
            reject("ERROR_PREPARE_TOKEN_INVOICE", error.localizedDescription, error)
        }
    }

    @objc(finalizeTokenInvoice:resolve:reject:)
    func rn_finalizeTokenInvoice(_ params: [String: Any],
                                 resolve: @escaping RCTPromiseResolveBlock,
                                 reject: @escaping RCTPromiseRejectBlock) {
        do {
            guard let receiverIdentityPublicKeyArray = params["receiverIdentityPublicKey"] as? [Any],
                  let receiverIdentityPublicKey = arrayToData(receiverIdentityPublicKeyArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid receiverIdentityPublicKey format"])
            }
            guard let network = params["network"] as? Int else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid network format"])
            }
            guard let fieldsBytesArray = params["sparkInvoiceFieldsBytes"] as? [Any],
                  let sparkInvoiceFieldsBytes = arrayToData(fieldsBytesArray) else {
                throw NSError(domain: "", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid sparkInvoiceFieldsBytes format"])
            }
            let signature = optionalArrayToData(params["signature"])

            let request = FinalizeTokenInvoiceRequest(
                receiverIdentityPublicKey: receiverIdentityPublicKey,
                network: UInt32(network),
                sparkInvoiceFieldsBytes: sparkInvoiceFieldsBytes,
                signature: signature
            )

            let result = try finalizeTokenInvoice(request: request)
            resolve(result)
        } catch {
            reject("ERROR_FINALIZE_TOKEN_INVOICE", error.localizedDescription, error)
        }
    }

    func constantsToExport() -> [AnyHashable : Any]! {
        return [:]
    }
}

private extension Array where Element == UInt8 {
    var data: Data {
        return Data(self)
    }
}

File: packages/spark-sdk/android/src/main/java/com/sparktokenprimitives/SparkTokenPrimitivesModule.kt
package com.sparktokenprimitives

import com.facebook.react.bridge.*
import com.facebook.react.module.annotations.ReactModule
import uniffi.spark_token_primitives.*

@ReactModule(name = SparkTokenPrimitivesModule.NAME)
class SparkTokenPrimitivesModule(reactContext: ReactApplicationContext) : ReactContextBaseJavaModule(reactContext) {
    companion object {
        const val NAME = "SparkTokenPrimitivesModule"
    }

    override fun getName(): String = NAME

    private fun ReadableArray.toByteArray(): ByteArray {
        return this.toArrayList().map { (it as Number).toByte() }.toByteArray()
    }

    private fun ByteArray.toWritableArray(): WritableArray {
        val array = Arguments.createArray()
        this.forEach { array.pushInt(it.toInt()) }
        return array
    }

    private fun ReadableMap.optionalByteArray(key: String): ByteArray? {
        if (!hasKey(key) || isNull(key)) return null
        return getArray(key)?.toByteArray()
    }

    private fun parseSelectedOutput(map: ReadableMap): SelectedTokenOutput {
        val previousTransactionHash = map.getArray("previousTransactionHash")?.toByteArray()
            ?: throw Exception("Invalid previousTransactionHash format")
        val previousTransactionVout = map.getInt("previousTransactionVout").toUInt()
        val ownerPublicKey = map.getArray("ownerPublicKey")?.toByteArray()
            ?: throw Exception("Invalid ownerPublicKey format")
        val tokenIdentifier = map.getArray("tokenIdentifier")?.toByteArray()
            ?: throw Exception("Invalid tokenIdentifier format")
        val tokenAmount = map.getArray("tokenAmount")?.toByteArray()
            ?: throw Exception("Invalid tokenAmount format")
        return SelectedTokenOutput(
            previousTransactionHash = previousTransactionHash,
            previousTransactionVout = previousTransactionVout,
            ownerPublicKey = ownerPublicKey,
            tokenIdentifier = tokenIdentifier,
            tokenAmount = tokenAmount
        )
    }

    private fun parseReceiverOutput(map: ReadableMap): ReceiverTokenOutput {
        val receiverSparkAddress = map.getString("receiverSparkAddress")
            ?: throw Exception("Invalid receiverSparkAddress format")
        val tokenIdentifier = map.optionalByteArray("tokenIdentifier")
        val tokenAmount = map.optionalByteArray("tokenAmount")
        return ReceiverTokenOutput(
            receiverSparkAddress = receiverSparkAddress,
            tokenIdentifier = tokenIdentifier,
            tokenAmount = tokenAmount
        )
    }

    private fun parseSignatureWithIndex(map: ReadableMap): SignatureWithIndexInput {
        val inputIndex = map.getInt("inputIndex").toUInt()
        val publicKey = map.getArray("publicKey")?.toByteArray()
            ?: throw Exception("Invalid publicKey format")
        val signature = map.getArray("signature")?.toByteArray()
            ?: throw Exception("Invalid signature format")
        return SignatureWithIndexInput(
            inputIndex = inputIndex,
            publicKey = publicKey,
            signature = signature
        )
    }

    @ReactMethod
    fun constructPartialTransferTransaction(params: ReadableMap, promise: Promise) {
        try {
            val identityPublicKey = params.getArray("identityPublicKey")?.toByteArray()
                ?: throw Exception("Invalid identityPublicKey format")

            val selectedOutputsArray = params.getArray("selectedOutputs")
                ?: throw Exception("Invalid selectedOutputs format")
            val selectedOutputs = mutableListOf<SelectedTokenOutput>()
            for (i in 0 until selectedOutputsArray.size()) {
                val map = selectedOutputsArray.getMap(i)
                    ?: throw Exception("Invalid selectedOutput at index $i")
                selectedOutputs.add(parseSelectedOutput(map))
            }

            val receiverOutputsArray = params.getArray("receiverOutputs")
                ?: throw Exception("Invalid receiverOutputs format")
            val receiverOutputs = mutableListOf<ReceiverTokenOutput>()
            for (i in 0 until receiverOutputsArray.size()) {
                val map = receiverOutputsArray.getMap(i)
                    ?: throw Exception("Invalid receiverOutput at index $i")
                receiverOutputs.add(parseReceiverOutput(map))
            }

            val operatorKeysArray = params.getArray("operatorIdentityPublicKeys")
                ?: throw Exception("Invalid operatorIdentityPublicKeys format")
            val operatorIdentityPublicKeys = mutableListOf<ByteArray>()
            for (i in 0 until operatorKeysArray.size()) {
                val keyArray = operatorKeysArray.getArray(i)
                    ?: throw Exception("Invalid operator key at index $i")
                operatorIdentityPublicKeys.add(keyArray.toByteArray())
            }

            val network = params.getInt("network").toUInt()
            val validityDurationSeconds = params.getDouble("validityDurationSeconds").toLong().toULong()
            val clientCreatedTimestampUnixMicros = params.getDouble("clientCreatedTimestampUnixMicros").toLong()
            val withdrawBondSats = params.getDouble("withdrawBondSats").toLong().toULong()
            val withdrawRelativeBlockLocktime = params.getDouble("withdrawRelativeBlockLocktime").toLong().toULong()
            val executeBeforeUnixMicros: Long? =
                if (params.hasKey("executeBeforeUnixMicros") && !params.isNull("executeBeforeUnixMicros"))
                    params.getDouble("executeBeforeUnixMicros").toLong()
                else null

            val request = TransferBuildRequest(
                identityPublicKey = identityPublicKey,
                selectedOutputs = selectedOutputs,
                receiverOutputs = receiverOutputs,
                operatorIdentityPublicKeys = operatorIdentityPublicKeys,
                network = network,
                validityDurationSeconds = validityDurationSeconds,
                clientCreatedTimestampUnixMicros = clientCreatedTimestampUnixMicros,
                withdrawBondSats = withdrawBondSats,
                withdrawRelativeBlockLocktime = withdrawRelativeBlockLocktime,
                executeBeforeUnixMicros = executeBeforeUnixMicros
            )

            val result = uniffi.spark_token_primitives.constructPartialTransferTransaction(request)

            val resultMap = Arguments.createMap().apply {
                putArray("partialTokenTransactionBytes", result.partialTokenTransactionBytes.toWritableArray())
                putArray("partialTokenTransactionHash", result.partialTokenTransactionHash.toWritableArray())
            }
            promise.resolve(resultMap)
        } catch (e: Exception) {
            promise.reject("ERROR_CONSTRUCT_PARTIAL_TRANSFER_TRANSACTION", e)
        }
    }

    @ReactMethod
    fun hashPartialTokenTransaction(params: ReadableMap, promise: Promise) {
        try {
            val bytes = params.getArray("partialTokenTransactionBytes")?.toByteArray()
                ?: throw Exception("Invalid partialTokenTransactionBytes format")
            val result = uniffi.spark_token_primitives.hashPartialTokenTransaction(bytes)
            promise.resolve(result.toWritableArray())
        } catch (e: Exception) {
            promise.reject("ERROR_HASH_PARTIAL_TOKEN_TRANSACTION", e)
        }
    }

    @ReactMethod
    fun buildBroadcastTransactionRequest(params: ReadableMap, promise: Promise) {
        try {
            val identityPublicKey = params.getArray("identityPublicKey")?.toByteArray()
                ?: throw Exception("Invalid identityPublicKey format")
            val partialTokenTransactionBytes = params.getArray("partialTokenTransactionBytes")?.toByteArray()
                ?: throw Exception("Invalid partialTokenTransactionBytes format")

            val signaturesArray = params.getArray("ownerSignatures")
                ?: throw Exception("Invalid ownerSignatures format")
            val ownerSignatures = mutableListOf<SignatureWithIndexInput>()
            for (i in 0 until signaturesArray.size()) {
                val map = signaturesArray.getMap(i)
                    ?: throw Exception("Invalid signature at index $i")
                ownerSignatures.add(parseSignatureWithIndex(map))
            }

            val request = BroadcastBuildRequest(
                identityPublicKey = identityPublicKey,
                partialTokenTransactionBytes = partialTokenTransactionBytes,
                ownerSignatures = ownerSignatures
            )

            val result = uniffi.spark_token_primitives.buildBroadcastTransactionRequest(request)
            promise.resolve(result.toWritableArray())
        } catch (e: Exception) {
            promise.reject("ERROR_BUILD_BROADCAST_TRANSACTION_REQUEST", e)
        }
    }

    @ReactMethod
    fun prepareTokenInvoice(params: ReadableMap, promise: Promise) {
        try {
            val receiverIdentityPublicKey = params.getArray("receiverIdentityPublicKey")?.toByteArray()
                ?: throw Exception("Invalid receiverIdentityPublicKey format")
            val network = params.getInt("network").toUInt()
            val tokenIdentifier = params.optionalByteArray("tokenIdentifier")
            val tokenAmount = params.optionalByteArray("tokenAmount")
            val memo = if (params.hasKey("memo") && !params.isNull("memo")) params.getString("memo") else null
            val senderSparkAddress = if (params.hasKey("senderSparkAddress") && !params.isNull("senderSparkAddress"))
                params.getString("senderSparkAddress") else null
            val expiryTimeUnixMillis: ULong? =
                if (params.hasKey("expiryTimeUnixMillis") && !params.isNull("expiryTimeUnixMillis"))
                    params.getDouble("expiryTimeUnixMillis").toLong().toULong()
                else null
            val invoiceId = params.optionalByteArray("invoiceId")

            val request = PrepareTokenInvoiceRequest(
                receiverIdentityPublicKey = receiverIdentityPublicKey,
                network = network,
                tokenIdentifier = tokenIdentifier,
                tokenAmount = tokenAmount,
                memo = memo,
                senderSparkAddress = senderSparkAddress,
                expiryTimeUnixMillis = expiryTimeUnixMillis,
                invoiceId = invoiceId
            )

            val result = uniffi.spark_token_primitives.prepareTokenInvoice(request)

            val resultMap = Arguments.createMap().apply {
                putArray("sparkInvoiceFieldsBytes", result.sparkInvoiceFieldsBytes.toWritableArray())
                putArray("sparkInvoiceHash", result.sparkInvoiceHash.toWritableArray())
                putString("unsignedSparkAddress", result.unsignedSparkAddress)
            }
            promise.resolve(resultMap)
        } catch (e: Exception) {
            promise.reject("ERROR_PREPARE_TOKEN_INVOICE", e)
        }
    }

    @ReactMethod
    fun finalizeTokenInvoice(params: ReadableMap, promise: Promise) {
        try {
            val receiverIdentityPublicKey = params.getArray("receiverIdentityPublicKey")?.toByteArray()
                ?: throw Exception("Invalid receiverIdentityPublicKey format")
            val network = params.getInt("network").toUInt()
            val sparkInvoiceFieldsBytes = params.getArray("sparkInvoiceFieldsBytes")?.toByteArray()
                ?: throw Exception("Invalid sparkInvoiceFieldsBytes format")
            val signature = params.optionalByteArray("signature")

            val request = FinalizeTokenInvoiceRequest(
                receiverIdentityPublicKey = receiverIdentityPublicKey,
                network = network,
                sparkInvoiceFieldsBytes = sparkInvoiceFieldsBytes,
                signature = signature
            )

            val result = uniffi.spark_token_primitives.finalizeTokenInvoice(request)
            promise.resolve(result)
        } catch (e: Exception) {
            promise.reject("ERROR_FINALIZE_TOKEN_INVOICE", e)
        }
    }
}
